presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -160,7 +172,7 @@ private synchronized boolean tryUpdateUserMemory(long delta)
             verify(future.isDone(), "future should be done");
             return true;
         }
-        if (queryMemoryContext.getUserMemory() + delta > maxMemory) {
+        if (queryMemoryContext.getUserMemory() + delta > maxUserMemory) {
```

Comment:
From your explanation above, this (calling `queryMemoryContext.getUserMemory()`) would lead to dead lock.